### DISCUSSION TITLE
ValueError: Unpack failed: incomplete input - when custom ner is loaded on Linux server. All works perfectly fine on my local host - windows 11

### DISCUSSION THREAD

**Discussion by rumnen** [explosion/spaCy#12584] [Help: Other Questions] [answered]
Similar issue was discussed before here: https://github.com/explosion/spaCy/discussions/10016

Despite that I receive same error `ValueError: Unpack failed: incomplete input`, I believe that this case is quite different as the error occurs only on Linux server and when custom ner is loaded.

I tried to install on the Linux server different versions and their related models. Same code, with the same method of installation, same versions of the package and model work perfectly fine on my PC.

`pip install spacy`
`python -m spacy download en_core_web_sm`

here are some OS details:
Linux kernel version 4.18.0-425.13.1.lve.el8.x86_64,
Apache version 2.4.57

I can post the entire log of the traceback plus my code if that would be helpful.

Thank you,
Rumen

**Comment by adrianeboyd** [comment, marked as answer]
Are you checking your model into a git repo at any point (which would be related to the issue you linked)?

What kind of platform/environment are you using to train and package your model? How are you copying it to the linux server?

The traceback+code might be be helpful. You can attach it as a `.txt` file if it gets too long.

**Reply by rumnen** [reply]
Thanks for replying. I trained the model in google colab and now it works. However, I think that resolving this issue is important and I  will answer all our questions.

I trained the model in anaconda environment but it worked perfectly fine in few other environments on my PC. Also I created on my PC an environment with exactly the same libraries as the environment I created on the Linux server and the model was working fine. I used to_disk() and spacy.load() to save and load the model. I exported the model by simply copying and pasting its folder.

Later I bumped into this article: https://github.com/explosion/spaCy/issues/8306 so I tried pickeling the model, but later couldn't unpickle it because of 
`NotImplementedError("cannot instantiate %r on your system"
NotImplementedError: cannot instantiate 'WindowsPath' on your system`
which by the way is not as easy to resolve issue as it may look, probably because of the sub directories within the model which would keep their windowsPath.
here is the log f the traceback:
[stderr.log](https://github.com/explosion/spaCy/files/11378702/stderr.log)
and the training code:


`
def train_spacy(data, num_epochs):

    training_data = data
    nlp = spacy.blank('en')
    if 'ner' not in nlp.pipe_names:
        ner = nlp.create_pipe('ner')
        nlp.add_pipe('ner', last=True)
    for _, annotations in training_data:
        for ent in annotations:
            ner.add_label(ent[2])
    other_pipes = [pipe for pipe in nlp.pipe_names if pipe != 'ner']
    with nlp.disable_pipes(*other_pipes):
        optimizer = nlp.begin_training()
        for epoch in range(num_epochs):
            print(f'starting epoch {epoch}')
            random.shuffle(training_data)
            losses = {}
            for text, annotations in training_data:
                doc = nlp.make_doc(text)
                example = Example.  from_dict(doc, annotations)
                nlp.update([example], drop=.2, sgd=optimizer, losses=losses)
            print(losses)
    return nlp
`

**Reply by adrianeboyd** [reply]
What program are you using when you are "copying and pasting its folder" to the linux server?

We wouldn't recommend using pickle for this. You should be able to use `nlp.to_disk` and `spacy package` to save models that work across multiple platforms, and possibly there's a similar issue to the git eol setting for the program you're copying with, and we can add a note to our docs if there's a particular setting you need to copy msgpack files successfully?

**Reply by rumnen** [reply]
I saved the model with `nlp.to_disk` and only used pickle when looking for alternatives ways to transfer it.

**Reply by adrianeboyd** [reply]
I suspect the problem is related to how it's being copied. How are you copying it?

**Reply by rumnen** [reply]
Copy and paste of the entire folder containing the model or I use FileZilla when uploading to server via ftp again the entire folder.

**Reply by adrianeboyd** [reply]
Can you try packaging the model with `spacy package`, copying the `.tar.gz` or `.whl` file from the `dist/` directory, and unpacking or installing that file on the server instead?

**Reply by adrianeboyd** [reply]
It's probably the ftp server that's modifying line endings in some files when it shouldn't, similar to the problem with git in the linked issue.
